# 108-2 Data Mining
## Homework #3 : Association Rule
## Instructor: Hsueh-Chan Lu
## Submitted by: Carlo D. Pastoral
## Student ID: N26087048

- In this homework, the task is to pick a multi-value attribute from my data and analyze the association relationships between those values.

- To do the task, I am going to use python as my programming language.
- To strat with, I am going to import the essential packages for exploring my datatset.

In [61]:
import pandas as pd 
import numpy as np
import json
import pprint
from scipy import stats
from collections import  Counter
from scipy.stats import entropy
pd.set_option('display.max_rows', None)


- Now lets print a 10 random sample from the daset and pick an attribute with multi-value.

In [62]:
data = pd.read_csv('Data/USvideos.csv')
sample = data.sample(n=50)
sample


37005  $1000 Survival Kit in a Case - https://youtu.be/3g5O-kT9m8k7 Ice Cream Gadgets put to the Test part 2 - https://youtu.be/IbJwUqKgHkQSubscribe to: 2nd channel - https://www.youtube.com/user/origami768instagram https://instagram.com/crazyrussianhacker/facebook - https://www.facebook.com/CrazyRussianHackerDON'T TRY THIS AT HOME!Business email: crh.inquire@gmail.comFAN MAIL:  CRAZY RUSSIAN HACKERP.O. Box 49Waynesville, NC 28786DISCLAIMER: In this video description contains affiliate links, which means that if you click on one of the product links, I’ll receive commission.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
2411   To get this complete recipe with instructions and measurements, check out my website: http://www.LauraintheKitchen.com\n\nInstagram: http://www.instagram.com/mrsvitale\n\nOfficial Facebook Page: http://www.facebook.com/LauraintheKitchen\n\nContact: Business@LauraintheKitchen.com\n\nTwitter: @Lauraskitchen                                                                                                                                                                                              

- From the 10 random samples, I saw that the "Tag" column is a multi-value attribute and it will be a good candidate for analyzing the association relationships between the tags. 
- However there is just one problem, **It is dirty.** 
- So first we need to clean this multi-value attribute first..

- To do this, first let us create a new dataframe container called **new_df** and put the 'video_id', 'tags', and 'views' cloumns here. 

In [63]:
new_df = pd.DataFrame(sample,columns = ['video_id','tags','views'])

 - Here I defined a funtion for cleaning every entry in **tags** column and converting each as list of tags while storing all in another list so I have a list of lists. Later on I will use this for applying Apriori.

 - I also changed the content of the **tags** column in the **new_df** dataframe with the cleaned one.

In [64]:
def cleanString(incomingString):
    newstring = incomingString
    newstring = newstring.replace("\"","")
    newstring = list(filter(None,newstring.split('|')))
    return newstring

tagsList = []
tagsList = [ cleanString(col['tags']) for idx,col in new_df.iterrows()] 
new_df['tags'] = tagsList
new_df


ke, keke palmer, Keke Palmer (Musical Artist), music, official music video, keke palmer music video, palmer, pregame, music video, new song, 2017, 2017 new song, new song video]                                                                                                                                                                                                                                                                                                                                        
21565  [Brian Hull, Brian Hull Impressions, Brian Hull Voices, Disney, Mickey Mouse, Minnnie Mouse, Mickey Impression, Minnie Impression, Disney Impressions, Disney Voices, Drive-Thru, Drive-Thru Pranks, Funny, Harwarming, Nice, Family Friendly]                                                                                                                                                                                                                                                                              
35576  [jacksepticeye, colour blind, colour blind test, color blind test, colour blindness, jacksepticeye colour blind, colour blind awareness, colour blind glasses, enchroma, seeing colour, colour test, colour number test, colour blind number test, color test, enchroma glasses, color blind glasses, Taking A Colour Blind Test With Colour Blind Glasses]                                                                                                                                                                 
6570   [Bicycles, bloomberg, business news, Cycling, Bike racing, custom bikes, vintage, hand-crafted, gifts, handmade, MADE, documentary, dorset, UK, England]                                                                                                                                                                                                                                                                                                                                                                    
36392  [family friendly, kids games, family games, challenges, competition, family fun, kid friendly, giant games, bottle flip, battle, trick shots, edgeland, giant, blind, boardgame, giant board game, team edge challenge, edge games, team edge, teamedge, matthias, team edge gaming]                                                                                                                                                                                                                                        
10732  [christmas, turkey, stuffing, recipe, food, cooking, cooking with your mouth, cooking on the frankston line, nathan ceddia, nowness, jamie oliver, how to make, turkey stuffing, gross, splosh, disgusting, shocking, yuck, raw food, feed me, spitting, chewing with your mouth open, sucking, licking, gaging, gage, spit, egg, onion, loud eating, asmr, rezept, food fetish, sexy lady, sexy, food porn, food tube, food hacks, food asmr, food blog, food blogger, food documentary, food diary, food design]          
28344  [humor, cyanide & happiness, explosm.net, shorts, funny, cyanide and happiness, safe neighborhood, cy&h, cartoon movies, C&H, cartoon, friends, neighbors, happiness, cyanide, explosm comics, c&h shorts, family, explosmentertainment, exlposm, neighborhood, c&hshorts, explosm, house, c and h shorts, satire, neighbor, explosm animated, animated, cyanide and happiness shorts, home, comedy]                                                                                                                        
39668  [Tia Mowry, Tia Mowry's Quick Fix, Tia Mowry Quick Fix, Quick Fix Tia Mowry, Quick Fix, hacks, recipes, diy, lifestyle how to, tips, tricks, Tia Mowry YouTube channel, Tia Mowry YouTube, baby girl nursery, nursery reveal, baby girl, second baby, baby number 2, nursery tour, nursery, baby girl nursery tour, nursery ideas, girl nursery tour, baby girl nursery decor, baby girl nursery ideas,

- Now that I have already prepared the data, I can now preceed in finding the association rules.
- To do that I will be using a python package called **apyori** to use its apriori function.
- Its **apriori** function only requires minimum values for **support**, **confidence**, and **lift**.
- It also allows the user to define a minimum number of words that will be scanned for antecedent and consequent.

In [65]:
from apyori import apriori

\begin{equation*}
Support(\{{{X}\}}\rightarrow\{{{Y}\}}) = \frac{Number\;of\;transactions\;containing\;both\;X\;and\;Y}{Total\;number\;of\;transactions}
\end{equation*}

\begin{equation*}
Confidence(\{{{X}\}}\rightarrow\{{{Y}\}}) = \frac{Number\;of\;transactions\;containing\;both\;X\;and\;Y}{Number\;of\;transactions\;containting X}
\end{equation*}

\begin{equation*}
Lift(\{{{X}\}}\rightarrow\{{{Y}\}}) = \frac{\frac{Number\;of\;transactions\;containing\;both\;X\;and\;Y}{Number\;of\;transactions\;containting\;X}}{Fraction\;of\;transactions\;containing\;Y}
\end{equation*}

- Now, to define a minimum values for **support**, **confidence**, and **lift** I will follow the formulas above. 
- I would like to set the minimum support to 2/50 so that I will get those Xs and Ys that occured twice over the fifty samples.
- For the confidence I will set it to ratio of 1/4. 
- For the minimum lift, I set it to 10 to limit the popularity of Ys  

In [71]:
association_rules = list(apriori(tagsList,min_support = 3/50, min_confidence = .25,min_lift = 10, min_length = 3))
len(association_rules)

16

- The function the I have used above returns a list of association rules. 
- And by getting the length of this list, I see that I've got a list of 16 association rules.
- To print the results I need to iterate through this list.

In [67]:

for item in association_rules:

    # first index of the inner list
    # Contains base item and add item
    pair = item[0] 
    items = [x for x in pair]
    print("Rule: " + items[0] + " -> " + items[1])

    #second index of the inner list
    print("Support: " + str(item[1]))

    #third index of the list located at 0th
    #of the third index of the inner list

    print("Confidence: " + str(item[2][0][2]))
    print("Lift: " + str(item[2][0][3]))
    print("=====================================")


Rule: family friendly -> bottle flip
Support: 0.06
Confidence: 1.0
Lift: 12.5
Rule: trick shots -> bottle flip
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: funny video -> celebrities
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: humor -> celebrities
Support: 0.06
Confidence: 1.0
Lift: 10.0
Rule: jokes -> celebrities
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: family friendly -> trick shots
Support: 0.06
Confidence: 0.75
Lift: 12.5
Rule: football -> soccer
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: humor -> funny video
Support: 0.06
Confidence: 1.0
Lift: 10.0
Rule: funny video -> jokes
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: humor -> jokes
Support: 0.06
Confidence: 0.6
Lift: 10.0
Rule: family friendly -> bottle flip
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: humor -> funny video
Support: 0.06
Confidence: 1.0
Lift: 16.666666666666668
Rule: funny video -> jokes
Support: 0.06
Confidence

- We can see that from these associatin rules, most are pretty obvious such **football -> soccer**. 
- We can also notice that most of the confidence value are equal to one but we can also notice that those association rule that has confidence value of less than 1,   the relativeness of antecedent and consequent is questionable. 
- Let us see for example the **Rule: family friendly -> **trick shots**, it has a support value of 0.06 which means that it **family friendly** appeared three times as with the **trick shots**  in the data. 
- And, it has a confidence value of 0.75 which means that it **trick shots** tag is more likely to use together with **family friendly** but, we intuitively know that these two tags have a weak association and there might be misleading about its confidence value. 
- So, to verify the association between these two tags we are going to look at the **lift** value. 
- We can see that the lift value for this association rule is **12.5**. Since the value is greatly more than 1, now this verifies that having a **family friendly** tag has a high chance of also having the **trick shots** tag in the video post.  